gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/GValue.java

```diff
@@ -0,0 +1,446 @@
+
+    private GValue(final GType type, final V value) {
+        this(null, type, value);
+    }
+
+    private GValue(final String name, final GType type, final V value) {
+        if (name != null && name.startsWith("_")) {
+            throw new IllegalArgumentException(String.format("Invalid GValue name [%s]. Should not start with _.", name));
+        }
+        this.name = name;
+        this.type = type;
+        this.value = value;
+    }
+
+    /**
+     * Determines if the value held by this object was defined as a variable or a literal value. Literal values simply
+     * have no name.
+     */
+    public boolean isVariable() {
+        return this.name != null;
+    }
+
+    /**
+     * Gets the name of the variable if it was defined as such and returns empty if the value was a literal.
+     */
+    public String getName() {
+        return this.name;
+    }
+
+    /**
+     * Gets the type of the value. The explicit type could be determined with {@code instanceof} on the value, but this
+     * might be helpful for cases where the value was constructed with a {@code null} value which might just return as
+     * {@code Object}.
+     */
+    public GType getType() {
+        return this.type;
+    }
+
+    /**
+     * Determines if the value held is of a {@code null} value.
+     */
+    public boolean isNull() {
+        return this.value == null;
+    }
+
+    /**
+     * Gets the value.
+     */
+    public V get() {
+        return this.value;
+    }
+
+    @Override
+    public String toString() {
+        return isVariable() ?
+                String.format("%s&%s", name, value) : Objects.toString(value);
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        final GValue<?> gValue = (GValue<?>) o;
+        return Objects.equals(name, gValue.name) && type == gValue.type && Objects.equals(value, gValue.value);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(name, type, value);
+    }
+
+    /**
+     * Create a new {@code Var} from a particular value but without the specified name. If the argument provide is
+     * already a {@code GValue} then it is returned as-is.
+     *
+     * @param value the value of the variable
+     */
+    static <V> GValue<V> of(final V value) {
+        if (value instanceof GValue) return (GValue) value;
+        return new GValue<>(GType.getType(value), value);
+    }
+
+    /**
+     * Create a new {@code Var} with the specified name and value.. If the argument provide is  already a
+     * {@code GValue} then it is returned as-is.
+     *
+     * @param name the name of the variable
+     * @param value the value of the variable
+     */
+    public static <V> GValue<V> of(final String name, final V value) {
+        if (value instanceof GValue) throw new IllegalArgumentException("value cannot be a GValue");
```

Comment:
This logic contradicts the javadoc and is also inconsistent with `static <V> GValue<V> of(final V value)`.

Reply:
Returning a `GValue` as is may not be desirable for this method as the users intentions are unclear in any case where the newly provided name does not match the existing name in the GValue. I have updated the javadoc accordingly but kept the implemented behavior.